iree/hal/metal/metal_buffer.h

```diff
@@ -0,0 +1,103 @@
+      device_size_t allocation_size, device_size_t byte_offset,
+      device_size_t byte_length, id<MTLBuffer> buffer,
+      id<MTLCommandQueue> transfer_queue);
+
+  // Creates a MetalBuffer instance without retaining the given id<MTLBuffer>.
+  static StatusOr<ref_ptr<MetalBuffer>> CreateUnretained(
+      MetalDirectAllocator* allocator, MemoryTypeBitfield memory_type,
+      MemoryAccessBitfield allowed_access, BufferUsageBitfield usage,
+      device_size_t allocation_size, device_size_t byte_offset,
+      device_size_t byte_length, id<MTLBuffer> buffer,
+      id<MTLCommandQueue> transfer_queue);
+
+  ~MetalBuffer() override;
+
+  id<MTLBuffer> handle() const { return metal_handle_; }
+
+ private:
+  // Creates a MetalBuffer instance without retaining the given id<MTLBuffer>.
+  MetalBuffer(MetalDirectAllocator* allocator, MemoryTypeBitfield memory_type,
+              MemoryAccessBitfield allowed_access, BufferUsageBitfield usage,
+              device_size_t allocation_size, device_size_t byte_offset,
+              device_size_t byte_length, id<MTLBuffer> buffer,
+              id<MTLCommandQueue> transfer_queue);
+
+  Status FillImpl(device_size_t byte_offset, device_size_t byte_length,
+                  const void* pattern, device_size_t pattern_length) override;
+  Status ReadDataImpl(device_size_t source_offset, void* data,
+                      device_size_t data_length) override;
+  Status WriteDataImpl(device_size_t target_offset, const void* data,
+                       device_size_t data_length) override;
+  Status CopyDataImpl(device_size_t target_offset, Buffer* source_buffer,
+                      device_size_t source_offset,
+                      device_size_t data_length) override;
+
+  Status MapMemoryImpl(MappingMode mapping_mode,
+                       MemoryAccessBitfield memory_access,
+                       device_size_t local_byte_offset,
+                       device_size_t local_byte_length,
+                       void** out_data) override;
+  Status UnmapMemoryImpl(device_size_t local_byte_offset,
+                         device_size_t local_byte_length, void* data) override;
+  Status InvalidateMappedMemoryImpl(device_size_t local_byte_offset,
+                                    device_size_t local_byte_length) override;
+  Status FlushMappedMemoryImpl(device_size_t local_byte_offset,
+                               device_size_t local_byte_length) override;
+
+  // Returns true if we need to automatically invaliate/flush CPU caches to keep
+  // memory hierarchy consistent.
+  //
+  // Note: this is needed when the buffer is requested with
+  // MemoryType::kHostCoherent bit but under the hood we are using memory types
+  // that does not have that property natively, e.g., MTLStorageModeManaged.
+  // Under such circumstances, we need to perform the invalidate/flush operation
+  // "automatically" for users.
+  bool NeedToAutoSynchronize() const;
```

Comment:
style nit: simple const accessors should be named like field accessors (`requires_autosync()` or something that you'd name a boolean flag)